File: mcp_compressor/just_bash_transform.py
"""AST-level pipe/redirection detection for just-bash custom commands.

Subclasses :class:`just_bash.Bash` to walk the parsed AST and inject
``MCP_TOONIFY=true|false`` env-var prefixes onto each invocation of one of
our wrapper commands, based on whether its stdout is piped (``cmd | jq``)
or redirected (``cmd > out.json``).  The wrapper command in
:mod:`mcp_compressor.bash_commands` reads ``ctx.env.MCP_TOONIFY`` to choose
between TOON and raw JSON output.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import replace
from typing import Any

from just_bash import Bash, ExecResult
from just_bash.ast.types import (
    AssignmentNode,
    CaseItemNode,
    CaseNode,
    CStyleForNode,
    ForNode,
    FunctionDefNode,
    GroupNode,
    IfClause,
    IfNode,
    LiteralPart,
    PipelineNode,
    ScriptNode,
    SimpleCommandNode,
    StatementNode,
    SubshellNode,
    UntilNode,
    WhileNode,
    WordNode,
)

#: Env var injected into wrapper invocations: ``"true"`` -> emit TOON,
#: ``"false"`` -> emit raw JSON.
MCP_TOONIFY_ENV_VAR = "MCP_TOONIFY"

#: Redirection operators that send stdout somewhere other than the caller.
#: ``2>&1`` is intentionally excluded (it dup's stderr, not stdout).
_OUTPUT_REDIR_OPERATORS = frozenset({">", ">>", ">|", "&>", "&>>", "<>"})


def resolve_toonify_from_ctx(env: Any, default: bool = False) -> bool | None:
    """Read ``MCP_TOONIFY`` from *env*, returning *default* if absent."""
    if env is None:
        return default
    try:
        raw = env.get(MCP_TOONIFY_ENV_VAR)
    except AttributeError:
        return default
    if raw is None:
        return default
    return raw.lower() == "true"


def _simple_command_name(cmd: SimpleCommandNode) -> str | None:
    """Return *cmd*'s literal name, or ``None`` for dynamic invocations."""
    if cmd.name is None:
        return None
    pieces: list[str] = []
    for part in cmd.name.parts:
        if isinstance(part, LiteralPart):
            pieces.append(part.value)
        else:
            return None
    return "".join(pieces) or None


def _has_output_redirection(cmd: SimpleCommandNode) -> bool:
    """Return ``True`` if *cmd* redirects fd 1 (stdout) somewhere else."""
    for redir in cmd.redirections or ():
        # ``fd`` defaults to 1 for ``>``/``>>``/``>|``; ``&>``/``&>>`` always cover stdout.
        if redir.operator in _OUTPUT_REDIR_OPERATORS and (
            redir.fd is None or redir.fd == 1 or redir.operator in {"&>", "&>>"}
        ):
            return True
    return False


def _make_env_assignment(name: str, value: str) -> AssignmentNode:
    """Build an :class:`AssignmentNode` for ``NAME=value``."""
    word = WordNode(parts=(LiteralPart(value=value),))
    return AssignmentNode(name=name, value=word)


def _inject_toonify_env(cmd: SimpleCommandNode, *, toonify: bool) -> SimpleCommandNode:
    """Prepend ``MCP_TOONIFY=...`` to *cmd*'s assignments (preserving user-set values)."""
    if any(a.name == MCP_TOONIFY_ENV_VAR for a in cmd.assignments or ()):
        return cmd
    new_assign = _make_env_assignment(MCP_TOONIFY_ENV_VAR, "true" if toonify else "false")
    return replace(cmd, assignments=(new_assign, *(cmd.assignments or ())))


def _transform_pipeline(pipeline: PipelineNode, custom_command_names: frozenset[str]) -> PipelineNode:
    """Inject toonify hints into wrapper commands in *pipeline*.

    A command is "piped" if it is not the last in its pipeline or has an
    output redirection of its own.
    """
    n = len(pipeline.commands)
    new_commands: list[Any] = []
    for index, cmd in enumerate(pipeline.commands):
        if isinstance(cmd, SimpleCommandNode):
            name = _simple_command_name(cmd)
            if name in custom_command_names:
                is_last = index == n - 1
                is_piped = (not is_last) or _has_output_redirection(cmd)
                cmd = _inject_toonify_env(cmd, toonify=not is_piped)
        else:
            cmd = _transform_compound(cmd, custom_command_names)
        new_commands.append(cmd)
    return replace(pipeline, commands=tuple(new_commands))


def _transform_statements(
    statements: Iterable[StatementNode], custom_command_names: frozenset[str]
) -> tuple[StatementNode, ...]:
    out: list[StatementNode] = []
    for stmt in statements:
        new_pipelines = tuple(_transform_pipeline(p, custom_command_names) for p in stmt.pipelines)
        out.append(replace(stmt, pipelines=new_pipelines))
    return tuple(out)


def _transform_compound(node: Any, custom_command_names: frozenset[str]) -> Any:
    """Recurse into compound nodes (``if``/``for``/``while``/subshell/etc.)."""
    if isinstance(node, IfNode):
        new_clauses = tuple(
            IfClause(
                condition=_transform_statements(c.condition, custom_command_names),
                body=_transform_statements(c.body, custom_command_names),
            )
            for c in node.clauses
        )
        new_else = _transform_statements(node.else_body, custom_command_names) if node.else_body is not None else None
        return replace(node, clauses=new_clauses, else_body=new_else)
    if isinstance(node, ForNode):
        return replace(node, body=_transform_statements(node.body, custom_command_names))
    if isinstance(node, CStyleForNode):
        return replace(node, body=_transform_statements(node.body, custom_command_names))
    if isinstance(node, (WhileNode, UntilNode)):
        return replace(
            node,
            condition=_transform_statements(node.condition, custom_command_names),
            body=_transform_statements(node.body, custom_command_names),
        )
    if isinstance(node, CaseNode):
        new_items = tuple(
            CaseItemNode(
                patterns=item.patterns,
                body=_transform_statements(item.body, custom_command_names),
            )
            for item in node.items
        )
        return replace(node, items=new_items)
    if isinstance(node, (SubshellNode, GroupNode)):
        return replace(node, body=_transform_statements(node.body, custom_command_names))
    if isinstance(node, FunctionDefNode) and node.body is not None:
        return replace(node, body=_transform_compound(node.body, custom_command_names))
    return node


def transform_script(ast: ScriptNode, custom_command_names: Iterable[str]) -> ScriptNode:
    """Rewrite *ast* to inject ``MCP_TOONIFY`` prefixes into wrapper commands."""
    names = frozenset(custom_command_names)
    if not names:
        return ast
    return replace(ast, statements=_transform_statements(ast.statements, names))


class _PipingAwareBash(Bash):
    """:class:`Bash` subclass that transforms each script's AST before execution.

    Wraps the underlying :class:`Interpreter`'s ``execute_script`` so we get a
    hook between parse and execute *without* duplicating ``Bash.exec``'s
    parse / env / cwd / error-handling logic.
    """

    def __init__(self, *args: Any, custom_command_names: Iterable[str], **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)
        self._mcp_custom_command_names: frozenset[str] = frozenset(custom_command_names)

        names = self._mcp_custom_command_names
        original_execute_script = self._interpreter.execute_script

        async def execute_script(node: Any) -> ExecResult:
            return await original_execute_script(transform_script(node, names))

        self._interpreter.execute_script = execute_script  # type: ignore[method-assign]


def build_piping_aware_bash(
    custom_commands: list[Any],
    *,
    fs: Any | None = None,
    cwd: str = "/",
    **kwargs: Any,
) -> Bash:
    """Construct a :class:`Bash` with built-ins + *custom_commands* + piping auto-detect."""
    from just_bash.commands import create_command_registry

    registry = create_command_registry()
    names: list[str] = []
    for command in custom_commands:
        registry[command.name] = command
        names.append(command.name)

    return _PipingAwareBash(
        custom_command_names=names,
        commands=registry,
        fs=fs,
        cwd=cwd,
        **kwargs,
    )

File: typescript/src/just_bash_transform.ts
/**
 * AST-level pipe/redirection detection for just-bash custom commands.
 *
 * Registers a ``TransformPlugin`` that walks the parsed AST and injects
 * ``MCP_TOONIFY=true|false`` env-var prefixes onto each invocation of one
 * of our wrapper commands, based on whether its stdout is piped
 * (``cmd | jq``) or redirected (``cmd > out.json``).  The wrapper command
 * in ``bash_commands.ts`` reads ``ctx.env.MCP_TOONIFY`` to choose between
 * TOON and raw JSON output.  User-supplied ``MCP_TOONIFY=...`` prefixes
 * are preserved.
 */

import type { Bash } from "just-bash";

/** Env var injected into wrapper invocations: "true" -> TOON, "false" -> JSON. */
export const MCP_TOONIFY_ENV_VAR = "MCP_TOONIFY";

/** Redirection operators that move stdout off the caller's pipe. */
const OUTPUT_REDIR_OPERATORS = new Set<string>([">", ">>", ">|", "&>", "&>>", "<>"]);

// --- minimal AST shape ------------------------------------------------------
// Type only the fields we touch so we don't depend on every AST detail.

interface LiteralPart {
  type: "Literal";
  value: string;
}

interface WordPart {
  type: string;
  value?: string;
}

interface WordNode {
  type: "Word";
  parts: WordPart[];
}

interface AssignmentNode {
  type: "Assignment";
  name: string;
  value: WordNode | null;
  append: boolean;
  array: WordNode[] | null;
}

interface RedirectionNode {
  type: "Redirection";
  fd: number | null;
  fdVariable?: string;
  operator: string;
  target: unknown;
}

interface SimpleCommandNode {
  type: "SimpleCommand";
  assignments: AssignmentNode[];
  name: WordNode | null;
  args: WordNode[];
  redirections: RedirectionNode[];
  line?: number;
}

interface PipelineNode {
  type: "Pipeline";
  commands: AnyCommandNode[];
  // ... other fields we don't touch
}

interface StatementNode {
  type: "Statement";
  pipelines: PipelineNode[];
  // ... other fields we don't touch
}

interface ScriptNode {
  type: "Script";
  statements: StatementNode[];
}

// All compound nodes have either ``body`` or ``clauses``/``items`` of
// ``StatementNode[]`` that we recurse into.  We keep this loosely typed.
type AnyCommandNode = SimpleCommandNode | Record<string, unknown>;

interface TransformInput {
  ast: ScriptNode;
  metadata: Record<string, unknown>;
}

interface TransformOutput {
  ast: ScriptNode;
  metadata?: Record<string, unknown>;
}

interface TransformPlugin {
  transform(input: TransformInput): TransformOutput;
}

// --- helpers ----------------------------------------------------------------

function isLiteralPart(part: WordPart): part is LiteralPart {
  return part.type === "Literal" && typeof part.value === "string";
}

/** Return *cmd*'s literal name, or ``null`` for dynamic invocations. */
function simpleCommandName(cmd: SimpleCommandNode): string | null {
  if (cmd.name === null) {
    return null;
  }
  const pieces: string[] = [];
  for (const part of cmd.name.parts) {
    if (!isLiteralPart(part)) {
      return null;
    }
    pieces.push(part.value);
  }
  const joined = pieces.join("");
  return joined.length > 0 ? joined : null;
}

/** Return ``true`` if *cmd* redirects fd 1 (stdout) somewhere else. */
function hasStdoutRedirection(cmd: SimpleCommandNode): boolean {
  for (const redir of cmd.redirections ?? []) {
    if (!OUTPUT_REDIR_OPERATORS.has(redir.operator)) {
      continue;
    }
    // fd defaults to 1 for >/>>/>|; &>/&>> always cover stdout.
    if (
      redir.fd === null ||
      redir.fd === 1 ||
      redir.operator === "&>" ||
      redir.operator === "&>>"
    ) {
      return true;
    }
  }
  return false;
}

function makeAssignment(name: string, value: string): AssignmentNode {
  return {
    type: "Assignment",
    name,
    value: { type: "Word", parts: [{ type: "Literal", value }] },
    append: false,
    array: null,
  };
}

function isSimpleCommand(node: unknown): node is SimpleCommandNode {
  return (
    typeof node === "object" &&
    node !== null &&
    (node as { type?: string }).type === "SimpleCommand"
  );
}

function injectToonify(cmd: SimpleCommandNode, toonify: boolean): void {
  // Preserve user-set MCP_TOONIFY assignments.
  for (const assignment of cmd.assignments ?? []) {
    if (assignment.name === MCP_TOONIFY_ENV_VAR) {
      return;
    }
  }
  cmd.assignments = [
    makeAssignment(MCP_TOONIFY_ENV_VAR, toonify ? "true" : "false"),
    ...(cmd.assignments ?? []),
  ];
}

function transformPipeline(pipeline: PipelineNode, names: Set<string>): void {
  const n = pipeline.commands.length;
  pipeline.commands.forEach((cmd, index) => {
    if (isSimpleCommand(cmd)) {
      const name = simpleCommandName(cmd);
      if (name !== null && names.has(name)) {
        const isLast = index === n - 1;
        const isPiped = !isLast || hasStdoutRedirection(cmd);
        injectToonify(cmd, !isPiped);
      }
    } else {
      transformCompound(cmd, names);
    }
  });
}

function transformStatements(statements: StatementNode[], names: Set<string>): void {
  for (const stmt of statements) {
    for (const pipeline of stmt.pipelines) {
      transformPipeline(pipeline, names);
    }
  }
}

/** Recurse into compound nodes (``if``/``for``/``while``/subshell/etc.) by shape. */
function transformCompound(node: Record<string, unknown>, names: Set<string>): void {
  for (const value of Object.values(node)) {
    if (Array.isArray(value)) {
      for (const item of value) {
        if (item && typeof item === "object") {
          const t = (item as { type?: string }).type;
          if (t === "Statement") {
            transformStatements(value as StatementNode[], names);
            break;
          }
          if (
            "body" in item ||
            "condition" in item ||
            "patterns" in item ||
            "clauses" in item ||
            "elseBody" in item
          ) {
            transformCompound(item as Record<string, unknown>, names);
          }
        }
      }
    } else if (value && typeof value === "object") {
      transformCompound(value as Record<string, unknown>, names);
    }
  }
}

// --- public plugin factory --------------------------------------------------

/** Build a ``TransformPlugin`` that injects ``MCP_TOONIFY`` prefixes (in-place AST mutation). */
export function createPipingHintPlugin(customCommandNames: readonly string[]): TransformPlugin {
  const names = new Set(customCommandNames);
  return {
    transform({ ast, metadata }) {
      if (names.size === 0) {
        return { ast, metadata };
      }
      transformStatements(ast.statements, names);
      return { ast, metadata };
    },
  };
}

/** Read ``MCP_TOONIFY`` from *env*, returning *defaultValue* if absent/invalid. */
export function resolveToonifyFromEnv(
  env: Map<string, string> | Record<string, string> | undefined,
  defaultValue: boolean,
): boolean {
  if (!env) {
    return defaultValue;
  }
  let raw: string | undefined;
  if (env instanceof Map) {
    raw = env.get(MCP_TOONIFY_ENV_VAR);
  } else {
    raw = env[MCP_TOONIFY_ENV_VAR];
  }
  if (raw === undefined) {
    return defaultValue;
  }
  const normalized = raw.trim().toLowerCase();
  if (normalized === "true") {
    return true;
  }
  if (normalized === "false") {
    return false;
  }
  return defaultValue;
}

/** Register :func:`createPipingHintPlugin` on *bash*. */
export function installPipingHintPlugin(bash: Bash, customCommandNames: readonly string[]): void {
  bash.registerTransformPlugin(createPipingHintPlugin(customCommandNames));
}
